### PR TITLE
Fix editor routing for static hosting

### LINKED ISSUE #13: Fix routing by changing hosting
I don't think GitHub Pages can be configured to support routing beyond static sites. This causes issues with the level editor routing. Should probably move over to something else.

### PR DESCRIPTION
**Pull request by codingcampbell**
Fixes #13

- Adds a symlink of editor/index.html -> index.html
- Adds .nojekyll so gh-pages routes symlinks

Reference:

- Symlinks in gh-pages: https://github.com/Sidnicious/gh-pages-symlink-test
- ReduxRouter basename support: https://stackoverflow.com/a/33376415

### CODE REVIEW

**Comment by codingcampbell** [comment]
Surprisingly, this works. Demo: http://codingcampbell.github.io/i-have-to-return-some-videotapes/ 

Kinda hacky though, and perhaps some of those hacks (e.g. .nojekyll) should exist in `gh-pages` only. Speaking of which, the `gh-pages` branch also has to add `<base>` tag for assets to load :frowning:. See https://github.com/codingcampbell/i-have-to-return-some-videotapes/commit/910819aa

**Comment by stevenhauser** [comment]
Welllll look at this. I'll check this out later today / this week. I'm fine with a semi-hacky solution for this minor thing since it's pretty small on code.

File: src/utils/createStore.js [hunk 1]
```diff
@@ -17,6 +17,13 @@ import * as appReducer from 'state';
 
 import { initialState } from 'state/initialState';
 
+const createHistoryWithBasename = (historyOptions) => {
+  return useBasename(createHistory)({
+    basename: location.hostname === 'localhost' ? '/' : '/i-have-to-return-some-videotapes',
```
Comment (stevenhauser) [inline src/utils/createStore.js]: Could this use `location.pathname` instead of a ternary and hardcoded values?

**Comment by stevenhauser** [comment]
Regarding the `<base>` stuff, how have you been building `gh-pages`? My workflow has been:

```
# start on `master`
git branch -D gh-pages
git checkout -b gh-pages
npm run build
git add static
git push
```

I considered making all of this a script but haven't had the need yet. With `<base>` in `index.html` in `gh-pages` would we basically move to just pushing changes to `static/`? I've been doing a full branch off of `master` so far just because I've had larger changes outside of just JS code.

File: src/utils/createStore.js (hunk 1 above)
Comment (codingcampbell) [inline src/utils/createStore.js]: That doesn't handle the case of landing on `/i-have-to.../editor/` though. Also a better ternary would be like a webpack-provided one at build-time

**Comment by stevenhauser** [comment]
Oh one other issue I noticed on your fork is that going directly to `/editor` messes up the footer text and URL because of [this garbage](https://github.com/stevenhauser/i-have-to-return-some-videotapes/blob/master/src/components/App/App.js#L13).

**Comment by codingcampbell** [comment]
`how have you been building gh-pages?`

I'm webpack noob, so I just ran with `webpack -p` and it output `style/bundle.js`. I then manually add `base` to index and ship away.

The more I think about it, I think this should be implemented as a build script. Excluding the Router change, webpack could be generating everything else: the `index.html` with a base tag, the `editor/index.html` dupe, the `.jekyll`, etc. Basically things that _only_ need to exist for static hosting.

File: src/utils/createStore.js (hunk 1 above)
Comment (stevenhauser) [inline src/utils/createStore.js]: Ah good point on `/editor`. But that route could be a constant or something and is always relative, so we could potentially do `location.pathname.replace('/editor', '')` which feels pretty nasty, but does keep everything relative and not hardcoded and not require any additional webpack :corn:fig.

**Comment by stevenhauser** [comment]
Yeah I like that idea of that stuff only happening in a build script. But I don't want to start adding a bunch of webpack :corn:ditionals for things like the `<base>` tag, so I wonder if there's away around that. If that's all that would need adjusted though, I'd be on board with that.

**Comment by stevenhauser** [comment]
Here's what still needs to happen here:

- [ ] Have everything happen in a build script if we can do that pretty simply
- [ ] Clean up the ternary in `createStore` if possible
- [ ] Fix the footer issue [here](https://github.com/stevenhauser/i-have-to-return-some-videotapes/blob/master/src/components/App/App.js#L13)

@codingcampbell Do you want to :corn:tribute any of these? I'm fine with running with your idea here if you're not up for coding any of these adjustments—you've already contributed a great idea.

**Comment by codingcampbell** [comment]
@stevenhauser I'll take a look at a build script today. I really prefer outputting the `.nojekyll` and `index.html` clones into some build folder that can be shipped to a static site. Also, coincidentally, I have been ramping up on webpack during holiday break, so.. good timing

**Comment by stevenhauser** [comment]
Coolio, that works for me. Thanks!

**Comment by codingcampbell** [comment]
Ok @stevenhauser here is what I've come up with:

1. `server.js` is now a ~100loc build script called `tasks.js.` The `npm start` command works the same.
2. `npm run build` works the same too, but now uses `tasks.js` instead of invoking `webpack`. This is because the webpack config is conditional now. One interesting side-effect is that a user no longer needs to `install -g` anything to use any of this.
3. All assets have moved to the `static/` folder, which is also where we place build artifacts. This is to avoid copying things like audio files at build time.
4. `index.html` now resides in `src/templates/`. The build script essentially copies this into `static/` with the proper `<base>` tag filled out
5. The ternary basename logic has  been replaced with `webpack.DefinePlugin` inlining a string constant into the code. This is the same technique the React`NODE_ENV` dev-code conditional is using.
6. Since `npm run build` works the same, a new `npm run build-static` command has landed to take the extra steps needed to make static hosting work. This is where the `.nojekyll` file is generated and `index.html` copied for all possible routes

So to test this, I ran `npm run build-static` and committed the contents of `static/` to `gh-pages`. It looks like this: https://github.com/codingcampbell/i-have-to-return-some-videotapes/tree/gh-pages and hosts like you would expect.

I didn't bother with the `editor/` thing because it seems to be working fine. I think the router is taking basename into consideration when giving you `pathname` because the condition is not failing for me.

The drawbacks I see are just :lipstick: stuff I think. For example, the `i-have-to-return...` path is hard-coded for the `build-static` task :unamused: 

Oh hey, one more cool side effect: the build is now 374k (down from 419k) :sparkles: 
I think this is because `DefinePlugin` is better at propograting the `NODE_ENV=production` stuff. Or maybe it's just my uglify options..

Hope this is helpful. Don't feel obliged to merge this in because of time spent. Truly most time was spent reading the webpack docs :eyeglasses:

**Comment by stevenhauser** [comment]
Wheeewww this is a lot more complicated than I expected. I liked the initial solution you came up with because it was only a few LOC and thought we'd be able to just tweak it. I'll take a closer look at this over the next couple of days. Thanks for putting all this time and effort in!

**Comment by stevenhauser** [comment]
Closing since this was superseded by #15. Thanks for your idea and time @codingcampbell!

**Pull request closed without merging**